src/harness/fourslash.ts

```diff
@@ -828,48 +830,121 @@ namespace FourSlash {
-                for (const m of markerName) this.verifyCompletionsAt(m, expected, options);
-                return;
+        public verifyCompletions(options: FourSlashInterface.VerifyCompletionsOptions) {
+            if (options.at !== undefined) {
+                if (typeof options.at === "string") {
+                    this.goToMarker(options.at);
+                }
+                else {
+                    for (const a of options.at) this.verifyCompletions({ ...options, at: a });
+                    return;
+                }
             }
 
-            this.goToMarker(markerName);
-
-            const actualCompletions = this.getCompletionListAtCaret(options);
+            const actualCompletions = this.getCompletionListAtCaret({ ...options.preferences, triggerCharacter: options.triggerCharacter });
             if (!actualCompletions) {
-                if (expected === undefined) return;
+                if (options.are === undefined) return;
                 this.raiseError(`No completions at position '${this.currentCaretPosition}'.`);
             }
 
-            if (actualCompletions.isNewIdentifierLocation !== (options && options.isNewIdentifierLocation || false)) {
-                this.raiseError(`Expected 'isNewIdentifierLocation' to be ${options && options.isNewIdentifierLocation}, got ${actualCompletions.isNewIdentifierLocation}`);
+            if (actualCompletions.isNewIdentifierLocation !== (options.isNewIdentifierLocation || false)) {
+                this.raiseError(`Expected 'isNewIdentifierLocation' to be ${options.isNewIdentifierLocation || false}, got ${actualCompletions.isNewIdentifierLocation}`);
             }
 
-            const actual = actualCompletions.entries;
+            const actualByName = ts.createMap<ts.CompletionEntry>();
+            for (const entry of actualCompletions.entries) {
+                if (actualByName.has(entry.name)) {
+                    // TODO: GH#23587
+                    if (entry.name !== "undefined" && entry.name !== "require") this.raiseError(`Duplicate completions for ${entry.name}`);
+                }
+                else {
+                    actualByName.set(entry.name, entry);
+                }
+            }
 
+            if ("are" in options) {
+                if (options.are === undefined) this.raiseError("Expected no completions");
+                this.verifyCompletionsAreExactly(actualCompletions.entries, toArray(options.are));
+            }
+            else {
+                if (options.includes) {
+                    for (const include of toArray(options.includes)) {
+                        const name = typeof include === "string" ? include : include.name;
+                        const found = actualByName.get(name);
+                        if (!found) this.raiseError(`No completion ${name} found`);
+                        this.verifyCompletionEntry(found, include);
+                    }
+                }
+                else {

```

Comment:
Are `includes` and `excludes` mutually exclusive?

Reply:
No, good catch!